### PR TITLE
Fixed :class:`~.Vector` tip

### LINKED ISSUE #968: Wrong orientation of vector tip
## Description of bug / unexpected behavior

Orientation of tips of some vectors are wrong

## How to reproduce the issue
```py
class Vectors(ThreeDScene):
    def construct(self):
        self.add(Vector(RIGHT), Vector(UP), Vector(OUT))
        self.set_camera_orientation(phi=PI/3, theta=PI/6)
        self.begin_ambient_camera_rotation(rate=2)
        self.wait(PI)
```

## Additional media files
Vectors

### PR DESCRIPTION
**Pull request by abhi3019**
Angle of the :class:`~.Vector` tip off the x-y plane was not considered while positioning :class:`~.TipableVMobject` in 3D.

## Motivation

Fixes #968 

## Overview / Explanation for Changes

Currently, angle of the vector above the plane is not considered while positioning its tip which is the cause of #968 .https://github.com/ManimCommunity/manim/blob/8fce5a74cbd66631ee4bc1419e186d07cd5590c9/manim/mobject/geometry.py#L168

## Oneline Summary of Changes

```
- Fixed bug in vector tip orientation (:pr:`PR NUMBER HERE`)
```

## Testing Status

Working Example:
```py
class VectorsTipFix(ThreeDScene):
    def construct(self):
        import itertools
        # test vectors from all quadrants
        vectors = map(Vector, itertools.product([2, -2], repeat=3))
        self.add(
            NumberPlane().set_opacity(.25), 
            *vectors
        )
        self.set_camera_orientation(phi = PI/3, theta = -PI/2 - 0.2)
        self.begin_ambient_camera_rotation(rate=0.7)
        self.wait(8)
```

https://user-images.githubusercontent.com/73361366/106113555-a0fd7d80-6174-11eb-8398-f92ac5d926d4.mp4

Also added an appropriate unit test.
## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The oneline summary has been included [in the wiki](https://github.com/ManimCommunity/manim/wiki/Changelog-for-next-release)

### CODE REVIEW

**Comment by Pragyaan-6988** [comment]
The graphical unit tests of `ThreeDAxes` are failing, which is fine I guess?

**Comment by jsonvillanueva** [comment]
Thanks for your contribution/quick bug fix @Pragyaan-6988!

> The graphical unit test of `ThreeDAxes` is failing, which is fine I guess?

It's fine in that your implementation fixes the bug, but we still want all the tests to pass. The graphical test (in tests/test_graphical_units/test_threed.py) will have to be updated in this PR to pass on each OS since your implementation changes the output of this 3D video. I'm unsure why they're not failing for Linux/Mac OS and only Windows 10, but the issue still stands. Checkout the [github wiki on testing](https://github.com/ManimCommunity/manim/wiki/Testing#graphical-unit-test) and update the failing test to your video output

**Comment by Pragyaan-6988** [comment]
Thanks for the pointers @jsonvillanueva. I'm new to unit tests and I'm beginning to understand how it all works. I updated the control data for the failing test and the tests pass locally (Windows 10). Now the test is failing for Linux/Mac OS, which I have no idea why.

**Comment by jsonvillanueva** [comment]
> Thanks for the pointers @jsonvillanueva. I'm new to unit tests and I'm beginning to understand how it all works. I updated the control data for the failing test and the tests pass locally (Windows 10). Now the test is failing for Linux/Mac OS, which I have no idea why.

I've tested locally on Windows and confirmed it works. On Manjaro Linux, it does indeed fail locally with this image provided by `pytest -k test_threed --show_diff`:
AxesTest

**Comment by naveen521kk** [comment]
On Windows 7 test's are failing

```
E           AssertionError: The frames don't match. Axes has been modified.
E           Please ignore if it was intended.
E           First unmatched index is at [238 736]: [159 159 159 255] != [160 160
 160 255]

tests\utils\GraphicalUnitTester.py:140: AssertionError
```
Figure_1

File: manim/mobject/geometry.py [hunk 1]
```diff
@@ -166,6 +166,10 @@ def position_tip(self, tip, at_start=False):
             handle = self.get_last_handle()
             anchor = self.get_end()
         tip.rotate(angle_of_vector(handle - anchor) - PI - tip.tip_angle)
+        tip.rotate(
+            PI / 2 - np.arccos(normalize(anchor)[2]),
```
Comment (behackl) [inline manim/mobject/geometry.py:170]: Just for the sake of testing whether it is a problem with the difference: I'd like to try something like

```suggestion
            0 if anchor[2] == 0 else PI / 2 - np.arccos(normalize(anchor)[2]),
```

Comment (Pragyaan-6988) [reply manim/mobject/geometry.py]: > Just for the sake of testing whether it is a problem with the difference: I'd like to try something like

Still the same result unfortunately

**Comment by jsonvillanueva** [comment]
Any updates on this @Pragyaan-6988?
I know that in one of your more recent PRs there was a similar issue that was resolved #1092

**Comment by Pragyaan-6988** [comment]
> Any updates on this

Hi @jsonvillanueva, unfortunately there isn't much I can think of doing to get the tests passing on ubuntu and macos.


> know that in one of your more recent PRs there was a similar issue that was resolved #1092

I just tried shifting a line of code to a different place and it just worked somehow.

Shall I try swapping the [two rotations](https://github.com/ManimCommunity/manim/pull/972/files#diff-b0ce9f0e97010be3f81191f949e3321debc323226bd70ca3bbbbf66dfe4a3ffeR168-R172)?

**Comment by Pragyaan-6988** [comment]
windows ✅ 
ubuntu ✅ 
macos ❌

**Comment by jsonvillanueva** [comment]
Last time it didn't pass on Ubuntu though, so not sure what changed.

**Comment by Pragyaan-6988** [comment]
> Last time it didn't pass on Ubuntu though, so not sure what changed.

Yeah it's just so strange.

**Comment by jsonvillanueva** [comment]
I'm unsure what to do about this PR. I don't think a full redesign of Arrows (and TippableVMobject) is out of the question; but I can't help but wonder how the Mac CI differs from the Windows/Linux workflow. Perhaps we need some test cases for the tip of the arrow since that seems to be the point at which the pytest diff first differs.

**Comment by abhi3019** [comment]
> Perhaps we need some test cases for the tip of the arrow since that seems to be the point at which the pytest diff first differs.

If I understood your comment correctly, I've already written a [test case for arrow tip](https://github.com/ManimCommunity/manim/pull/972/files#diff-ba73b8d3fd55d0cad719d4a5175f2b75e7734d0c207b99a8a489614a493862f2R11-R20) . May be this test is passing because I rounded off the coordinates to 8 decimal places?

**Comment by behackl** [comment]
This is now obsolete, given that #1924 has been merged.

**Pull request closed without merging**